src/transformers/models/jukebox/modeling_jukebox.py

```diff
@@ -0,0 +1,2974 @@
+        hidden_states = self.c_proj(hidden_states)
+        hidden_states = self.dropout(hidden_states)
+        return hidden_states
+
+
+class JukeboxLayerNorm(FusedLayerNorm):
+    def __init__(self, normalized_shape, eps=1e-5, elementwise_affine=True):
+        super().__init__(normalized_shape, eps=eps, elementwise_affine=elementwise_affine)
+        self.width = np.prod(normalized_shape)
+        self.max_numel = 65535 * self.width
+
+    def forward(self, input):
+        if input.numel() > self.max_numel:
+            return F.layer_norm(input, self.normalized_shape, self.weight, self.bias, self.eps).type_as(input)
+        else:
+            return super(JukeboxLayerNorm, self).forward(input).type_as(input)
+
+
+class JukeboxAttention(nn.Module):
+    def __init__(
+        self,
+        width,
+        n_ctx,
+        hidden_dim,
+        num_heads,
+        attn_dropout=0.0,
+        resid_dropout=0.0,
+        scale=True,
+        mask=False,
+        zero_out=False,
+        init_scale=1.0,
+        attn_func=0,
+        blocks=None,
+        spread=None,
+        encoder_dims=None,
+        lyric_enc_len=None,
+    ):
+        super().__init__()
+        self.width = width  # should have a better name
+        self.n_ctx = n_ctx  # NOTE: n_ctx could be different within operations. This is complete n_ctx
+        self.hidden_dim = hidden_dim
+        self.num_heads = num_heads
+        self.scale = scale
+        self.mask = mask
+        if attn_func == 6:
+            self.c_attn = JukeboxConv1D(width, hidden_dim)
+            self.c_enc_kv = JukeboxConv1D(width, hidden_dim * 2)
+        else:
+            self.c_attn = JukeboxConv1D(width, hidden_dim * 3)
+        self.c_proj = JukeboxConv1D(hidden_dim, width, zero_out)
+        self.attn_dropout = nn.Dropout(attn_dropout) if attn_dropout > 0.0 else lambda x: x
+        self.resid_dropout = nn.Dropout(resid_dropout) if resid_dropout > 0.0 else lambda x: x
+
+        # Sequence of length seq_len is factored as [blocks, seq_len // blocks]
+        self.attn_func = attn_func
+        if attn_func == 6:
+            self.qkv = self.decode_qkv
+        elif attn_func == 7:
+            self.qkv = self.prime_qkv
+        else:
+            self.qkv = self.factored_qkv
+
+        self.attn, self.attn_mask = {
+            0: (self.dense_attn, "autoregressive"),  # Attend to all positions
+            1: (self.block_attn, "autoregressive"),  # Attend to your block
+            2: (self.transpose_block_attn, "autoregressive"),  # Attend to transpose block
+            3: (self.prev_block_attn, None),  # Attend to previous block
+            4: (self.summary_attn, "summary"),  # Attend to last position of each block
+            5: (self.summary_spread_attn, "summary"),
+            6: (self.dense_attn, None),
+            7: (self.prime_attn, "prime"),
+        }[
+            attn_func
+        ]  # Attend to last key position of each block
```

Comment:
Absolutely not. Please use string code or an enum for the attention types and not an int like this.